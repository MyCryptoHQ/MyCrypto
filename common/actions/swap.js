// @flow
import type {
  OriginKindSwapAction,
  DestinationKindSwapAction,
  OriginAmountSwapAction,
  DestinationAmountSwapAction,
  LoadBityRatesSucceededSwapAction,
  DestinationAddressSwapAction,
  BityOrderCreateSucceededSwapAction,
  BityOrderCreateRequestedSwapAction,
  OrderStatusSucceededSwapAction,
  ChangeStepSwapAction,
  Pairs,
  RestartSwapAction,
  LoadBityRatesRequestedSwapAction,
  StopLoadBityRatesSwapAction,
  BityOrderResponse,
  BityOrderPostResponse,
  OrderStatusRequestedSwapAction,
  StopOrderTimerSwapAction,
  StartOrderTimerSwapAction,
  StartPollBityOrderStatusAction,
  StopPollBityOrderStatusAction
} from './swapTypes';

export function changeStepSwap(value: number): ChangeStepSwapAction {
  return {
    type: 'SWAP_STEP',
    value
  };
}

export function originKindSwap(value: string): OriginKindSwapAction {
  return {
    type: 'SWAP_ORIGIN_KIND',
    value
  };
}

export function destinationKindSwap(value: string): DestinationKindSwapAction {
  return {
    type: 'SWAP_DESTINATION_KIND',
    value
  };
}

export function originAmountSwap(value: ?number): OriginAmountSwapAction {
  return {
    type: 'SWAP_ORIGIN_AMOUNT',
    value
  };
}

export function destinationAmountSwap(
  value: ?number
): DestinationAmountSwapAction {
  return {
    type: 'SWAP_DESTINATION_AMOUNT',
    value
  };
}

export function loadBityRatesSucceededSwap(
  value: Pairs
): LoadBityRatesSucceededSwapAction {
  return {
    type: 'SWAP_LOAD_BITY_RATES_SUCCEEDED',
    value
  };
}

export function destinationAddressSwap(
  value: ?string
): DestinationAddressSwapAction {
  return {
    type: 'SWAP_DESTINATION_ADDRESS',
    value
  };
}

export function restartSwap(): RestartSwapAction {
  return {
    type: 'SWAP_RESTART'
  };
}

export function loadBityRatesRequestedSwap(): LoadBityRatesRequestedSwapAction {
  return {
    type: 'SWAP_LOAD_BITY_RATES_REQUESTED'
  };
}

export function stopLoadBityRatesSwap(): StopLoadBityRatesSwapAction {
  return {
    type: 'SWAP_STOP_LOAD_BITY_RATES'
  };
}

export function orderTimeSwap(value: number) {
  return {
    type: 'SWAP_ORDER_TIME',
    value
  };
}

export function bityOrderCreateSucceededSwap(
  payload: BityOrderPostResponse
): BityOrderCreateSucceededSwapAction {
  return {
    type: 'SWAP_BITY_ORDER_CREATE_SUCCEEDED',
    payload
  };
}

export function bityOrderCreateRequestedSwap(
  amount: number,
  destinationAddress: string,
  pair: string,
  mode: number = 0
): BityOrderCreateRequestedSwapAction {
  return {
    type: 'SWAP_ORDER_CREATE_REQUESTED',
    payload: {
      amount,
      destinationAddress,
      pair,
      mode
    }
  };
}

export function orderStatusSucceededSwap(
  payload: BityOrderResponse
): OrderStatusSucceededSwapAction {
  return {
    type: 'SWAP_BITY_ORDER_STATUS_SUCCEEDED',
    payload
  };
}

export function orderStatusRequestedSwap(): OrderStatusRequestedSwapAction {
  return {
    type: 'SWAP_BITY_ORDER_STATUS_REQUESTED'
  };
}

export function startOrderTimerSwap(): StartOrderTimerSwapAction {
  return {
    type: 'SWAP_ORDER_START_TIMER'
  };
}

export function stopOrderTimerSwap(): StopOrderTimerSwapAction {
  return {
    type: 'SWAP_ORDER_STOP_TIMER'
  };
}

export function startPollBityOrderStatus(): StartPollBityOrderStatusAction {
  return {
    type: 'SWAP_START_POLL_BITY_ORDER_STATUS'
  };
}

export function stopPollBityOrderStatus(): StopPollBityOrderStatusAction {
  return {
    type: 'SWAP_STOP_POLL_BITY_ORDER_STATUS'
  };
}
