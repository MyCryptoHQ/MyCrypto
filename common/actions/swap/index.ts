export * from './actionTypes';
export * from './actionCreators';
