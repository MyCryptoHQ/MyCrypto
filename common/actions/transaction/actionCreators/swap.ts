import {
  SwapEtherToTokenAction,
  SwapTokenToEtherAction,
  SwapTokenToTokenAction
} from '../actionTypes';
import { TypeKeys } from '../constants';

type TSwapTokenToEther = typeof swapTokenToEther;
const swapTokenToEther = (payload: SwapTokenToEtherAction['payload']): SwapTokenToEtherAction => ({
  type: TypeKeys.TOKEN_TO_ETHER_SWAP,
  payload
});

type TSwapEtherToToken = typeof swapEtherToToken;
const swapEtherToToken = (payload: SwapEtherToTokenAction['payload']): SwapEtherToTokenAction => ({
  payload,
  type: TypeKeys.ETHER_TO_TOKEN_SWAP
});

type TSwapTokenToToken = typeof swapTokenToToken;
const swapTokenToToken = (payload: SwapTokenToTokenAction['payload']): SwapTokenToTokenAction => ({
  payload,
  type: TypeKeys.TOKEN_TO_TOKEN_SWAP
});

export {
  swapEtherToToken,
  swapTokenToEther,
  swapTokenToToken,
  TSwapTokenToEther,
  TSwapEtherToToken,
  TSwapTokenToToken
};
