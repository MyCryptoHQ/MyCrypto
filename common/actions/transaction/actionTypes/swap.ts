import { TypeKeys } from 'actions/transaction/constants';
import {
  SetToFieldAction,
  SetValueFieldAction,
  SetTokenToMetaAction,
  SetTokenValueMetaAction,
  SetDataFieldAction
} from 'actions/transaction';

/* Swapping actions */
interface SwapTokenToEtherAction {
  type: TypeKeys.TOKEN_TO_ETHER_SWAP;
  payload: {
    to: SetToFieldAction['payload'];
    value: SetValueFieldAction['payload'];
    decimal: number;
  };
}
interface SwapEtherToTokenAction {
  type: TypeKeys.ETHER_TO_TOKEN_SWAP;
  payload: {
    to: SetToFieldAction['payload'];
    data: SetDataFieldAction['payload'];
    tokenTo: SetTokenToMetaAction['payload'];
    tokenValue: SetTokenValueMetaAction['payload'];
    decimal: number;
  };
}
interface SwapTokenToTokenAction {
  type: TypeKeys.TOKEN_TO_TOKEN_SWAP;
  payload: {
    to: SetToFieldAction['payload'];
    data: SetDataFieldAction['payload'];
    tokenValue: SetTokenValueMetaAction['payload'];
    decimal: number;
  };
}
type SwapAction = SwapEtherToTokenAction | SwapTokenToEtherAction | SwapTokenToTokenAction;

export { SwapTokenToEtherAction, SwapEtherToTokenAction, SwapAction, SwapTokenToTokenAction };
