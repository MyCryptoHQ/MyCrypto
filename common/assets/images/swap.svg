<?xml version="1.0" encoding="UTF-8"?>
<svg width="620px" height="620px" viewBox="0 0 620 620" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <!-- Generator: Sketch 48.2 (47327) - http://www.bohemiancoding.com/sketch -->
    <title>swap</title>
    <desc>Created with Sketch.</desc>
    <defs>
        <circle id="path-1" cx="170" cy="170" r="170"></circle>
        <circle id="path-3" cx="170" cy="170" r="170"></circle>
    </defs>
    <g id="Page-1" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <g id="swap">
            <g id="Yellow-Coin" transform="translate(0.000000, 280.000000)">
                <mask id="mask-2" fill="white">
                    <use xlink:href="#path-1"></use>
                </mask>
                <circle stroke="#0E97C0" stroke-width="20" cx="170" cy="170" r="160"></circle>
                <rect id="Rectangle" fill="#FFE14D" mask="url(#mask-2)" x="-1.13333333" y="1.13333333" width="173.4" height="340"></rect>
                <rect id="Rectangle" fill="#FFCC33" mask="url(#mask-2)" x="171.133333" y="1.13333333" width="173.4" height="340"></rect>
                <circle id="Oval-2" stroke="#333333" stroke-width="20" mask="url(#mask-2)" cx="170" cy="170" r="160"></circle>
                <circle id="Oval-3" stroke="#F28618" stroke-width="20" mask="url(#mask-2)" cx="170" cy="170" r="96.3333333"></circle>
            </g>
            <g id="Blue-Coin" transform="translate(280.000000, 0.000000)">
                <mask id="mask-4" fill="white">
                    <use xlink:href="#path-3"></use>
                </mask>
                <circle stroke="#0E97C0" stroke-width="20" cx="170" cy="170" r="160"></circle>
                <rect id="Rectangle" fill="#6EA6E8" mask="url(#mask-4)" x="-1.13333333" y="1.13333333" width="173.4" height="340"></rect>
                <rect id="Rectangle" fill="#5C9BE4" mask="url(#mask-4)" x="171.133333" y="1.13333333" width="173.4" height="340"></rect>
                <circle id="Oval-2" stroke="#333333" stroke-width="20" mask="url(#mask-4)" cx="170" cy="170" r="160"></circle>
                <circle id="Oval-3" stroke="#2F79CF" stroke-width="20" mask="url(#mask-4)" cx="170" cy="170" r="96.3333333"></circle>
            </g>
            <g id="Group" transform="translate(226.000000, 320.000000)" fill-rule="nonzero">
                <polygon id="Shape" fill="#0492BE" points="311.67 8 385.42 8 385.42 171.31 144.71 171.31 144.71 234.53 13 134.44 144.71 34.34 144.71 97.56 311.67 97.56"></polygon>
                <polygon id="Shape" fill="#103957" opacity="0.2" points="348.54 134.44 13 134.44 144.71 234.53 144.71 171.31 385.42 171.31 385.42 8 348.54 8"></polygon>
                <path d="M152.66,250.36 L0,134.36 L152.66,18.36 L152.66,89.61 L303.82,89.61 L303.82,0 L393.38,0 L393.38,179.11 L152.66,179.11 L152.66,250.36 Z M26.11,134.36 L136.85,218.52 L136.85,163.31 L377.58,163.31 L377.58,15.8 L319.63,15.8 L319.63,105.36 L136.86,105.36 L136.86,50.17 L26.11,134.36 Z" id="Shape" fill="#000000"></path>
            </g>
            <g id="Group" transform="translate(197.000000, 174.500000) rotate(180.000000) translate(-197.000000, -174.500000) translate(0.000000, 49.000000)" fill-rule="nonzero">
                <polygon id="Shape" fill="#0492BE" points="311.67 8 385.42 8 385.42 171.31 144.71 171.31 144.71 234.53 13 134.44 144.71 34.34 144.71 97.56 311.67 97.56"></polygon>
                <polygon id="Shape" fill="#103957" opacity="0.2" points="348.54 134.44 13 134.44 144.71 234.53 144.71 171.31 385.42 171.31 385.42 8 348.54 8"></polygon>
                <path d="M152.66,250.36 L0,134.36 L152.66,18.36 L152.66,89.61 L303.82,89.61 L303.82,0 L393.38,0 L393.38,179.11 L152.66,179.11 L152.66,250.36 Z M26.11,134.36 L136.85,218.52 L136.85,163.31 L377.58,163.31 L377.58,15.8 L319.63,15.8 L319.63,105.36 L136.86,105.36 L136.86,50.17 L26.11,134.36 Z" id="Shape" fill="#000000"></path>
            </g>
        </g>
    </g>
</svg>