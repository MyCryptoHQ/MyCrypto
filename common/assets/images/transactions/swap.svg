<svg width="45" height="45" viewBox="0 0 45 45" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M0 12.2727C0 5.49468 5.49468 0 12.2727 0H32.7273C39.5053 0 45 5.49468 45 12.2727V32.7273C45 39.5053 39.5053 45 32.7273 45H12.2727C5.49468 45 0 39.5053 0 32.7273V12.2727Z" fill="#FFD166"/>
<path d="M23.5686 26.3219C23.6386 25.8862 23.6733 25.4452 23.672 25.0034C23.672 23.411 23.1998 21.8543 22.3151 20.5302C21.4304 19.2062 20.1729 18.1742 18.7017 17.5648C17.2305 16.9554 15.6116 16.7959 14.0497 17.1066C12.4879 17.4173 11.0532 18.1841 9.92721 19.3101C8.80119 20.4362 8.03435 21.8708 7.72368 23.4327C7.41301 24.9945 7.57246 26.6134 8.18186 28.0846C8.79126 29.5559 9.82325 30.8133 11.1473 31.698C12.3535 32.504 13.7526 32.9676 15.196 33.0438" stroke="white" stroke-width="1.7" stroke-linecap="round" stroke-linejoin="round"/>
<path d="M25.8496 13.1918V8.72039C25.8485 8.58445 25.8057 8.45213 25.7269 8.34135C25.6481 8.23057 25.5372 8.14667 25.4091 8.10102C25.2811 8.05537 25.1421 8.05016 25.011 8.08611C24.8799 8.12206 24.763 8.19742 24.6761 8.302L22.3086 11.0981C16.1055 7.93676 6.77763 12.7738 8.48123 21.3351" stroke="white" stroke-width="1.7" stroke-linecap="round" stroke-linejoin="round"/>
<path d="M14.7898 27.0606L15.5739 38.6201C15.5733 38.7554 15.6141 38.8877 15.6908 38.9992C15.7676 39.1106 15.8765 39.196 16.0031 39.2437C16.1298 39.2915 16.2679 39.2994 16.3992 39.2663C16.5304 39.2333 16.6484 39.1609 16.7373 39.0589L19.0946 36.2832C27.5876 39.3735 34.2944 33.5703 33.3812 27.9766" stroke="white" stroke-width="1.7" stroke-linecap="round" stroke-linejoin="round"/>
<path d="M11.8234 25.5718C11.2216 21.3614 17.1391 19.6493 19.0325 23.2285" stroke="white" stroke-width="1.7" stroke-linecap="round" stroke-linejoin="round"/>
<path d="M33.3251 21.4339L32.6783 17.0273C32.6528 16.8721 32.5872 16.7263 32.4881 16.6041C32.389 16.482 32.2598 16.3877 32.1132 16.3307C31.9666 16.2737 31.8076 16.2559 31.652 16.279C31.4965 16.3021 31.3496 16.3654 31.2259 16.4625L28.4819 18.6182L27.7286 19.21C27.6062 19.3074 27.5116 19.4353 27.4541 19.5807C27.3966 19.7262 27.3783 19.8842 27.4009 20.039C27.4236 20.1937 27.4865 20.3398 27.5833 20.4627C27.6801 20.5855 27.8075 20.6808 27.9526 20.7391L28.8724 21.1045L32.0828 22.3802C32.2288 22.4439 32.3891 22.4674 32.5471 22.4483C32.7052 22.4293 32.8553 22.3684 32.9819 22.2719C33.1086 22.1754 33.2072 22.0469 33.2675 21.8995C33.3278 21.7522 33.3477 21.5915 33.3251 21.4339Z" stroke="white" stroke-width="1.7" stroke-linecap="round" stroke-linejoin="round"/>
<circle cx="30.8426" cy="19.628" r="8.06921" stroke="white" stroke-width="1.7" stroke-linecap="round" stroke-linejoin="round"/>
</svg>
