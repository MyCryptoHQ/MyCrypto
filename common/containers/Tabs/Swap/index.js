import React, { Component } from 'react';
import { connect } from 'react-redux';
import * as swapActions from 'actions/swap';
import * as swapTypes from 'actions/swapTypes';
import CurrencySwap from './components/CurrencySwap';
import CurrentRates from './components/CurrentRates';
import ReceivingAddress from './components/ReceivingAddress';
import SwapInfoHeader from './components/SwapInfoHeader';
import SwapProgress from './components/SwapProgress';

type ReduxStateProps = {
  step: string,
  destinationAddress: string,
  destinationKind: string,
  originKind: string,
  destinationKindOptions: String[],
  originKindOptions: String[],
  bityRates: boolean,
  originAmount: ?number,
  destinationAmount: ?number,
  // PART 3
  referenceNumber: string,
  timeRemaining: string,
  numberOfConfirmation: number,
  orderStep: number,
  orderStarted: boolean
};

type ReduxActionProps = {
  changeStepSwap: (value: number) => swapTypes.ChangeStepSwapAction,
  originKindSwap: (value: string) => swapTypes.OriginKindSwapAction,
  destinationKindSwap: (value: string) => swapTypes.DestinationKindSwapAction,
  originAmountSwap: (value: ?number) => swapTypes.OriginAmountSwapAction,
  destinationAmountSwap: (
    value: ?number
  ) => swapTypes.DestinationAmountSwapAction,
  loadBityRatesSwap: () => swapTypes.LoadBityRatesSwapAction,
  destinationAddressSwap: (
    value: ?string
  ) => swapTypes.DestinationAddressSwapAction,
  restartSwap: () => swapTypes.RestartSwapAction,
  stopLoadBityRatesSwap: () => swapTypes.StopLoadBityRatesSwapAction,
  // PART 3 (IGNORE FOR NOW)
  referenceNumberSwap: typeof swapActions.referenceNumberSwap
};

class Swap extends Component {
  props: ReduxActionProps & ReduxStateProps;

  componentDidMount() {
    this.props.loadBityRatesSwap();
  }

  componentWillUnmount() {
    this.props.stopLoadBityRatesSwap();
  }

  render() {
    let {
      // STATE
      bityRates,
      originAmount,
      destinationAmount,
      originKind,
      destinationKind,
      destinationKindOptions,
      originKindOptions,
      destinationAddress,
      step,
      referenceNumber,
      timeRemaining,
      numberOfConfirmations,
      orderStep,
      // ACTIONS
      restartSwap,
      stopLoadBityRatesSwap,
      changeStepSwap,
      originKindSwap,
      destinationKindSwap,
      originAmountSwap,
      destinationAmountSwap,
      destinationAddressSwap,
      referenceNumberSwap
    } = this.props;

    let ReceivingAddressProps = {
      destinationKind,
      destinationAddressSwap,
      destinationAddress,
      stopLoadBityRatesSwap,
      changeStepSwap,
      referenceNumberSwap
    };

    let SwapInfoHeaderProps = {
      referenceNumber,
      timeRemaining,
      originAmount,
      originKind,
      destinationKind,
      destinationAmount,
      restartSwap,
      numberOfConfirmations,
      orderStep
    };

    const { ETHBTC, ETHREP, BTCETH, BTCREP } = bityRates;

    const CurrentRatesProps = { ETHBTC, ETHREP, BTCETH, BTCREP };

    const CurrencySwapProps = {
      bityRates,
      originAmount,
      destinationAmount,
      originKind,
      destinationKind,
      destinationKindOptions,
      originKindOptions,
      originKindSwap,
      destinationKindSwap,
      originAmountSwap,
      destinationAmountSwap,
      changeStepSwap
    };

    return (
      <section className="container" style={{ minHeight: '50%' }}>
        <div className="tab-content">
          <main className="tab-pane swap-tab">
            {step === 1 &&
              <div>
                <CurrentRates {...CurrentRatesProps} />
                <CurrencySwap {...CurrencySwapProps} />
              </div>}
            {(step === 2 || step === 3) &&
              <SwapInfoHeader {...SwapInfoHeaderProps} />}
            {step === 2 && <ReceivingAddress {...ReceivingAddressProps} />}
            {step === 3 && <SwapProgress {...SwapInfoHeaderProps} />}
          </main>
        </div>
      </section>
    );
  }
}

function mapStateToProps(state) {
  return {
    step: state.swap.step,
    destinationAddress: state.swap.destinationAddress,
    originAmount: state.swap.originAmount,
    destinationAmount: state.swap.destinationAmount,
    originKind: state.swap.originKind,
    destinationKind: state.swap.destinationKind,
    destinationKindOptions: state.swap.destinationKindOptions,
    originKindOptions: state.swap.originKindOptions,
    bityRates: state.swap.bityRates,
    referenceNumber: state.swap.referenceNumber,
    timeRemaining: state.swap.timeRemaining,
    numberOfConfirmations: state.swap.numberOfConfirmations,
    orderStep: state.swap.orderStep,
    orderStarted: state.swap.orderStarted
  };
}

export default connect(mapStateToProps, swapActions)(Swap);
