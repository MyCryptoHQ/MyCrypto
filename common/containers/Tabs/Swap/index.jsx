import React, { Component } from 'react';
import { connect } from 'react-redux';
import { showNotification } from 'actions/notifications';
import * as swapActions from 'actions/swap';
import type {
  ChangeStepSwapAction,
  OriginKindSwapAction,
  DestinationKindSwapAction,
  OriginAmountSwapAction,
  DestinationAmountSwapAction,
  LoadBityRatesRequestedSwapAction,
  DestinationAddressSwapAction,
  RestartSwapAction,
  StopLoadBityRatesSwapAction,
  BityOrderCreateRequestedSwapAction,
  StartPollBityOrderStatusAction,
  StopOrderTimerSwapAction,
  StopPollBityOrderStatusAction
} from 'actions/swap';
import CurrencySwap from './components/CurrencySwap';
import CurrentRates from './components/CurrentRates';
import ReceivingAddress from './components/ReceivingAddress';
import SwapInfoHeader from './components/SwapInfoHeader';
import PartThree from './components/PartThree';

type ReduxStateProps = {
  step: string,
  destinationAddress: string,
  destinationKind: string,
  originKind: string,
  destinationKindOptions: String[],
  originKindOptions: String[],
  bityRates: {},
  originAmount: ?number,
  destinationAmount: ?number,
  isPostingOrder: boolean,
  isFetchingRates: boolean,
  bityOrder: {},
  secondsRemaining: ?number,
  paymentAddress: ?string,
  orderStatus: ?string,
  outputTx: ?string
};

type ReduxActionProps = {
  changeStepSwap: (value: number) => ChangeStepSwapAction,
  originKindSwap: (value: string) => OriginKindSwapAction,
  destinationKindSwap: (value: string) => DestinationKindSwapAction,
  originAmountSwap: (value: ?number) => OriginAmountSwapAction,
  destinationAmountSwap: (value: ?number) => DestinationAmountSwapAction,
  loadBityRatesRequestedSwap: () => LoadBityRatesRequestedSwapAction,
  destinationAddressSwap: (value: ?string) => DestinationAddressSwapAction,
  restartSwap: () => RestartSwapAction,
  stopLoadBityRatesSwap: () => StopLoadBityRatesSwapAction,
  bityOrderCreateRequestedSwap: (
    amount: number,
    destinationAddress: string,
    pair: string,
    mode: number
  ) => BityOrderCreateRequestedSwapAction,
  startPollBityOrderStatus: () => StartPollBityOrderStatusAction,
  stopOrderTimerSwap: () => StopOrderTimerSwapAction,
  stopPollBityOrderStatus: () => StopPollBityOrderStatusAction
};

class Swap extends Component {
  props: ReduxActionProps & ReduxStateProps;

  componentDidMount() {
    // TODO: Use `isFetchingRates` to show a loader
    this.props.loadBityRatesRequestedSwap();
  }

  componentWillUnmount() {
    this.props.stopLoadBityRatesSwap();
  }

  render() {
    let {
      // STATE
      bityRates,
      originAmount,
      destinationAmount,
      originKind,
      destinationKind,
      destinationKindOptions,
      originKindOptions,
      destinationAddress,
      step,
      bityOrder,
      secondsRemaining,
      paymentAddress,
      orderStatus,
      isPostingOrder,
      outputTx,
      // ACTIONS
      restartSwap,
      stopLoadBityRatesSwap,
      changeStepSwap,
      originKindSwap,
      destinationKindSwap,
      originAmountSwap,
      destinationAmountSwap,
      destinationAddressSwap,
      bityOrderCreateRequestedSwap,
      showNotification,
      startOrderTimerSwap,
      startPollBityOrderStatus,
      stopOrderTimerSwap,
      stopPollBityOrderStatus
    } = this.props;

    const { reference } = bityOrder;

    let ReceivingAddressProps = {
      isPostingOrder,
      originAmount,
      originKind,
      destinationKind,
      destinationAddressSwap,
      destinationAddress,
      stopLoadBityRatesSwap,
      changeStepSwap,
      bityOrderCreateRequestedSwap
    };

    let SwapInfoHeaderProps = {
      reference,
      secondsRemaining,
      originAmount,
      originKind,
      destinationKind,
      destinationAmount,
      restartSwap,
      orderStatus
    };

    const { ETHBTC, ETHREP, BTCETH, BTCREP } = bityRates;
    const CurrentRatesProps = { ETHBTC, ETHREP, BTCETH, BTCREP };

    const CurrencySwapProps = {
      showNotification,
      bityRates,
      originAmount,
      destinationAmount,
      originKind,
      destinationKind,
      destinationKindOptions,
      originKindOptions,
      originKindSwap,
      destinationKindSwap,
      originAmountSwap,
      destinationAmountSwap,
      changeStepSwap
    };

    const PaymentInfoProps = {
      originKind,
      originAmount,
      paymentAddress
    };

    const PartThreeProps = {
      ...SwapInfoHeaderProps,
      ...PaymentInfoProps,
      reference,
      startOrderTimerSwap,
      startPollBityOrderStatus,
      stopOrderTimerSwap,
      stopPollBityOrderStatus,
      showNotification,
      destinationAddress,
      outputTx
    };

    return (
      <section className="Tab-content swap-tab">
        {step === 1 && <CurrentRates {...CurrentRatesProps} />}
        {(step === 2 || step === 3) &&
          <SwapInfoHeader {...SwapInfoHeaderProps} />}

        <main className="Tab-content-pane">
          {step === 1 && <CurrencySwap {...CurrencySwapProps} />}
          {step === 2 && <ReceivingAddress {...ReceivingAddressProps} />}
          {step === 3 && <PartThree {...PartThreeProps} />}
        </main>
      </section>
    );
  }
}

function mapStateToProps(state) {
  return {
    outputTx: state.swap.outputTx,
    isPostingOrder: state.swap.isPostingOrder,
    orderStatus: state.swap.orderStatus,
    paymentAddress: state.swap.paymentAddress,
    step: state.swap.step,
    destinationAddress: state.swap.destinationAddress,
    originAmount: state.swap.originAmount,
    destinationAmount: state.swap.destinationAmount,
    originKind: state.swap.originKind,
    destinationKind: state.swap.destinationKind,
    destinationKindOptions: state.swap.destinationKindOptions,
    originKindOptions: state.swap.originKindOptions,
    bityRates: state.swap.bityRates,
    bityOrder: state.swap.bityOrder,
    secondsRemaining: state.swap.secondsRemaining,
    isFetchingRates: state.swap.isFetchingRates
  };
}

export default connect(mapStateToProps, { ...swapActions, showNotification })(
  Swap
);
