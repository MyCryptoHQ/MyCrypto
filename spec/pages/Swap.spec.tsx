import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Swap from 'containers/Tabs/Swap';
import shallowWithStore from '../utils/shallowWithStore';
import { createMockStore } from 'redux-test-utils';
import { INITIAL_STATE } from 'reducers/swap';

Enzyme.configure({ adapter: new Adapter() });

it('render snapshot', () => {
  const store = createMockStore({ swap: INITIAL_STATE });
  const component = shallowWithStore(<Swap />, store);

  expect(component).toMatchSnapshot();
});
