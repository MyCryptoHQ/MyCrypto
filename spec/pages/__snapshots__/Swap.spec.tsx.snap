// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render snapshot 1`] = `
<Swap
  bityOrder={Object {}}
  bityOrderCreateRequestedSwap={[Function]}
  bityOrderStatus={null}
  bityRates={
    Object {
      "allIds": Array [],
      "byId": Object {},
    }
  }
  changeStepSwap={[Function]}
  destination={
    Object {
      "amount": NaN,
      "id": "ETH",
    }
  }
  destinationAddress=""
  destinationAddressSwap={[Function]}
  history={
    Object {
      "action": "PUSH",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 2,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "e08jz7",
        "pathname": "/swap",
        "search": "",
        "state": Object {},
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  initSwap={[Function]}
  isFetchingRates={null}
  isOffline={false}
  isPostingOrder={false}
  loadBityRatesRequestedSwap={[Function]}
  loadShapeshiftRatesRequestedSwap={[Function]}
  location={
    Object {
      "hash": "",
      "key": "e08jz7",
      "pathname": "/swap",
      "search": "",
      "state": Object {},
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {},
      "path": "/swap",
      "url": "/swap",
    }
  }
  options={
    Object {
      "allIds": Array [],
      "byId": Object {},
    }
  }
  origin={
    Object {
      "amount": NaN,
      "id": "BTC",
    }
  }
  outputTx={null}
  paymentAddress={null}
  provider="shapeshift"
  restartSwap={[Function]}
  secondsRemaining={null}
  shapeshiftOrder={Object {}}
  shapeshiftOrderCreateRequestedSwap={[Function]}
  shapeshiftOrderStatus={null}
  shapeshiftRates={
    Object {
      "allIds": Array [],
      "byId": Object {},
    }
  }
  showNotification={[Function]}
  startOrderTimerSwap={[Function]}
  startPollBityOrderStatus={[Function]}
  startPollShapeshiftOrderStatus={[Function]}
  step={1}
  stopLoadBityRatesSwap={[Function]}
  stopLoadShapeshiftRatesSwap={[Function]}
  stopOrderTimerSwap={[Function]}
  stopPollBityOrderStatus={[Function]}
  stopPollShapeshiftOrderStatus={[Function]}
  swapProvider={[Function]}
/>
`;
